### PR TITLE
Allow unit-ful ScalarMappable data

### PR DESCRIPTION
**Pull request by dstansby**
## PR Summary
Allows 2D data with units attached to be used by `imshow`. Fixes https://github.com/matplotlib/matplotlib/issues/19476, and supersedes https://github.com/matplotlib/matplotlib/pull/19481.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -733,6 +735,31 @@ def set_data(self, A):
         self._rgbacache = None
         self.stale = True
 
+    def _convert_units(self, A):
+        # Take the first element since units expects a 1D sequence, not 2D
+        converter = munits.registry.get_converter(A[0])
```
Comment (tacaswell) [inline lib/matplotlib/image.py:740]: We probably want to stash the converter the first time through?

**Comment by tacaswell** [comment]
This required fewer changes than I had naively expected :)

File: lib/matplotlib/image.py (hunk 1 above)
Comment (dstansby) [reply lib/matplotlib/image.py:740]: Makes sense. It does raise the question of making converter immutable once it has been set once (ie. the first time `set_data()` is called). Should this be the case, or should we provide a method to manually change the converter after a `set_data()` call?

File: lib/matplotlib/colorbar.py [hunk 2]
```diff
@@ -470,9 +472,6 @@ def __init__(self, ax, mappable=None, *, cmap=None,
             linewidths=[0.5 * mpl.rcParams['axes.linewidth']])
         self.ax.add_collection(self.dividers)
 
-        self.locator = None
-        self.minorlocator = None
-        self.formatter = None
```
Comment (dstansby) [inline lib/matplotlib/colorbar.py:475]: These lines are redundant, as they're also run in `_reset_locator_formatter_scale()` below.

**Comment by dstansby** [comment]
Units are a quality of the mappable, not the Axis (or Axes) on which that mappable is plotted. When adding a colorbar with that mappable, the units of the mappable become associated with the colorbar Axis. In the same way that without units one can plot two datasets with different values that map to identical colours (e.g., using different norms), with this PR it is possible to do the same with different united data. If you want to tell the difference between these you can add two colorbars that will have the correct units attached.

> It introduces a significant inconsistency across the library with other methods that also map 2-D data.

It's inconsistent in the sense that other methods don't support 2D map data yet, but I would say that adding that support to those other methods is a feature request that shouldn't block this. This PR doesn't change current behaviour of those other methods. Do you think this is only worth going in if units are supported across **all** 2D plotting methods?

> I'm not against doing this, but would like to see a more general plan, or an explicit statement that we will not support other 2-D data types.

I'm not sure what you mean by 2-D data types here?

**Comment by jklymak** [comment]
I mean why would you prioritize `imshow` versus `pcolormesh` or `contour` (both of which I use _far_ more than `imshow`)

**Comment by dstansby** [comment]
🤷 no particular reason - in solar physics we use `imshow` a lot so I guess that's why. There's nothing in this PR stopping adding support for those other methods. If we agree on the general approach in this PR I can add support for those, either as additional commits here or in follow up PRs.

**Comment by story645** [comment]
> Units are a quality of the mappable, not the Axis (or Axes) on which that mappable is plotted. In the same way that without units one can plot two datasets with different values that map to identical colours (e.g., using different norms)

that's why I think units should really be stashed on the Norm (since that's what's doing the mappable->screen), but I feel like an implementation where the unit objects can decide which Norms they support (like categorical only works with NoNorm, dates w/ the non-log ones), or provide their own, is the one that leads to least mess.

**Comment by dstansby** [comment]
> that's why I think units should really be stashed on the Norm (since that's what's doing the mappable->screen)

I guess the units are eventually being stored on the colorbar Axis here. When plotting normally the Axis exists before you plot so it's possible to associate the converter + units with the Axis straight away. With colorbars the Axis doesn't exist until after plotting though, so we have to keep a copy of the converter + units on the mappable, then assign it to the colorbar Axis. This is fine because there is a  1 -> 1 mapping from colorbar to a mappable. 

> an implementation where the unit objects can decide which Norms they support (like categorical only works with NoNorm, dates w/ the non-log ones)

I had not thought of this point - is there a mechanism to do this currently (e.g., if I plot some dates and try and set a log scale does the date converter stop me doing that?)

> ETA: to be clear, I like this PR & think it's a great improvement, but I have the same concerns as Jody about inconsistent support. If this PR goes in, then there is unitized color, and I think that then should be folded back into not just the other 2D image methods, but also scatter.

👍  - I have targeted this to 3.6.0, will change it to draft status, and slowly add support for other methods (thanks for reminding me of scatter!) in subsequent commits.

**Comment by story645** [comment]
> (e.g., if I plot some dates and try and set a log scale does the date converter stop me doing that?)

I just realized that was a bad example since there are use cases where you maybe want log date, like paleoclimatology, but currently nope there' nothing preventing this & what happens is the scaling is applied to the floats that the convertor converts too.

**Comment by dstansby** [comment]
So it seems like the general solution also only requires a limited number of changes - notably adding support for units on the `ScalarMappable` class. Seems to work with contour(f), pcolor(mesh) and imshow so far, but I'll update this with some comprehensive tests shortly.

**Comment by dstansby** [comment]
I think this is good for review now. The API changes/what's new entries should explain everything that's going on, if they don't please leave a request for more docs!

**Comment by jklymak** [comment]
If I pass a mappable with units "mega-joules" and another to the same axes with units "giga-joules" and they use the same norm, what happens?  I would expect they would be converted to the same units, and plotted comparably.  I don't think that happens here...

**Comment by jklymak** [comment]
The point is that if they are not sharing the norm or colorbar, what is the advantage of dealing with the units versus just stripping the values out of the container?  Just formatting the colorbar?

**Comment by dstansby** [comment]
Yes, I think that's correct.

**Comment by jklymak** [comment]
Currently we use units to format axises for datetime and categorical.  Are there other uses?  I can _maybe_ see it being nice to format a colorbar with datetime, though I am skeptical.  What are the other use cases you envision for this?

**Comment by jklymak** [comment]
I am skeptical because units do not currently work as advertised for univariate axes, and here they only work in the restricted sense of formatting the axes with no design for unit conversion of the scalars.  I am also concerned because the new ability of colorbars to behave like other axes is very new, and this will definitely stress the new behaviour.  

If you look on Stackoverflow, unit conversion, particularly for categoricals (but datetime is a close second), is one of the leading sources of confusion.  I think we should be very cautious about adding to the unit API.

**Comment by dstansby** [comment]
> The original goal was that you could pass (104, 'cm') and (72, 'inches') to an axis, set its units to ('m') and the data would be plotted at 1.04 and 1.83. There is an example to this effect and it does not work.

It seems to work for me (albeit with an external units library):
```python
import astropy.units as u
from astropy.visualization import quantity_support
import numpy as np
import matplotlib.pyplot as plt
quantity_support()

x = np.arange(10)
y_rad = np.linspace(0, 2 * np.pi, 10) * u.rad
y_deg = np.linspace(0, 180, 10) * u.deg

fig, ax = plt.subplots()
ax.scatter(x, y_rad)
ax.scatter(x, y_deg)
# ax.xaxis.set_units(u.deg)  # This doesn't work
plt.show()
```

Changing the units does not though...

> I often have one colorbar _representing_ many mappables, usually in different axes. They don't "share" the colorbar in code, because we don't do that, but they do share the colorbar conceptually.

Right, but you have to manually make sure that all your data is comparable and have to *manually* ensure they're plotted with the same norm and colormap, and then add a colorbar from a single mappable, so something like:

```python
norm = Norm(vmin=vmin, vmax=vmax)
cmap = cmap

im1 = ax1.imshow(data1, norm=norm, cmap=cmap)
im2 = ax2.imshow(data2, norm=norm, cmap=cmap)
fig.colorbar(im2)
```
`im1` has nothing to do with the colorbar (in the code), but you just so happen to have scaled it so it makes sense with the colorbar. It's the same as plotting two sets of data on twinned Axes and then making sure that the scale and limits of both twinned Axes are the same.

As it stands with this PR the behaviour is exactly the same, ie. you would have to manually make sure multiple scalarmappables are scaled correctly for one colorbar. I think what you're getting at (please correct me if I'm wrong!) is having an API like:

```python
cbar = fig.colobar(norm=Norm(vmin, vmax), cmap=cmap)
im1 = ax1.imshow(data1, colobar=cbar)
im2 = ax2.imshow(data2, colobar=cbar)
```
and units/vmin/vmax being taken into account by the colorbar correctly? I think this would be an awesome API, and am +/-0 blocking this PR on such an API existing (if it's desirable).

So I'm happy to put this PR on hold, *but* I think it is a nice new feature, and want to be sure there's good reasons for putting it on hold.

**Comment by tacaswell** [comment]
This PR is affected by a re-writing of our history to remove a large number of accidentally committed files [see discourse](https://discourse.matplotlib.org/t/minor-re-writing-of-history-and-moving-from-master-to-main-for-default-branch/22354/6) for details.

To recover this PR it will need be rebased onto the new default branch (main).  There are several ways to accomplish this, but we recommend (assuming that you call the matplotlib/matplotlib remote `"upstream"`

```bash
git remote update
git checkout main
git merge --ff-only upstream/main
git checkout YOUR_BRANCH
git rebase --onto=main upstream/old_master
# git rebase -i main # if you prefer
git push --force-with-lease   # assuming you are tracking your branch
```

If you do not feel comfortable doing this or need any help please reach out to any of the Matplotlib developers.  We can either help you with the process or do it for you.

Thank you for your contributions to Matplotlib and sorry for the inconvenience.

**Comment by jklymak** [comment]
@dstansby let us know if/when you can make a meeting to discuss this.  If you can't make it, let us know and we can try and discuss without you

**Comment by jklymak** [comment]
@dstansby this needs a rebase so I moved to drafts....

**Comment by dstansby** [comment]
@jklymak thanks - I don't have any bandwidth to pick this back up and discuss it in the near future, so if anyone else wants to take it up feel free.

**Comment by jklymak** [comment]
We discussed this on the call last week, and should discuss again.  However, I will give my screed about categoricals:
[_Every single day_](https://stackoverflow.com/questions/70341043/why-plot-python-3-different-with-matlab-and-even-excel) on Stackoverflow there is at least one person who thinks Matplotlib is very broken because they plot 1000 numbers or dates, and get 1000 ticks because they mistakenly passed in strings for x and/or y.  _Every day_, often numerous times a day.  Then, instead of fixing their problem by converting to float or datetime, they try and mess around with the tick intervals using `set_xticks` or even [`set_major_locator`](https://stackoverflow.com/questions/70343472/entire-data-from-date-range-not-being-plotted-in-matplotlib), of course losing the proper spacing between the data, and churning out innumerate garbage. This is a huge usability problem that we have made by trying to be clever, because the user has no idea what is going on - I would argue that if someone is sophisticated enough to know about `set_major_locator` and they get caught by this, then we have made an API mistake.  

My point being that I _strongly_ feel people should have to manually opt in to categorical plotting.  I actually think they should have to manually opt into any unit handling, including dates (`ax.plot(x, y, xunits='dates')` is not too much to ask).  

For expanding categoricals into scalar mappables, I'm fine with the feature, but I think if we just do it via object inspection, folks will end up not understanding what has happened to their data, and will again think Matplotlib is broken.  In my opinion this API should be accompanied by some way of manually opting in.

**Pull request closed without merging**